### PR TITLE
Migrate the first half of the async demos

### PR DESCRIPTION
**Pull request by peter-csala**
# Description
- Migrated demos form 00 to 05 under Async folder 
- Renamed demo 00 to reflect V8's concepts
- Moved statistics fields from `SyncDemo` to `DemoBase`

### CODE REVIEW

File: PollyDemos/Async/AsyncDemo02_WaitAndRetryNTimes.cs [hunk 1]
```diff
@@ -1,102 +1,94 @@
...
-            var policy = Policy.Handle<Exception>().WaitAndRetryAsync(
-                3, // Retry 3 times
-                attempt => TimeSpan.FromMilliseconds(200), // Wait 200ms between each try.
-                (exception, calculatedWaitDuration) => // Capture some info for logging!
+            // Define our strategy:
+            var strategy = new ResiliencePipelineBuilder().AddRetry(new()
+            {
+                ShouldHandle = new PredicateBuilder().Handle<Exception>(),
+                MaxRetryAttempts = 3,
+                Delay = TimeSpan.FromMilliseconds(200),  // Wait 200ms between each try
```
Comment (martincostello) [inline PollyDemos/Async/AsyncDemo02_WaitAndRetryNTimes.cs:38]: Then we don't need to change two things if we changed the value of the delay. Feels a bit self-descriptive to me in any case.

```suggestion
                Delay = TimeSpan.FromMilliseconds(200),  // Wait between each try
```

File: PollyDemos/Async/AsyncDemo05_WaitAndRetryWithExponentialBackoff.cs [hunk 2]
```diff
@@ -1,112 +1,98 @@
...
-            using (var client = new HttpClient())
+            // Define our strategy:
+            var strategy = new ResiliencePipelineBuilder().AddRetry(new()
             {
-                totalRequests = 0;
-                var internalCancel = false;
-                // Do the following until a key is pressed
-                while (!internalCancel && !cancellationToken.IsCancellationRequested)
+                ShouldHandle = new PredicateBuilder().Handle<Exception>(),
+                MaxRetryAttempts = 6, // We could also retry indefinitely, but chose six times instead.
```
Comment (martincostello) [inline PollyDemos/Async/AsyncDemo05_WaitAndRetryWithExponentialBackoff.cs:41]:
```suggestion
                MaxRetryAttempts = 6, // We could also retry indefinitely by using {whatever the value to do that is}
```

File: PollyDemos/DemoBase.cs [hunk 3]
```diff
@@ -4,6 +4,11 @@ namespace PollyDemos
 {
     public abstract class DemoBase
     {
+        protected int totalRequests;
+        protected int eventualSuccesses;
+        protected int eventualFailures;
+        protected int retries;
```
Comment (martincostello) [inline PollyDemos/DemoBase.cs:10]: Let's make these properties with `PascalCase` naming.

**Review by martincostello — COMMENTED** [review]
Just a few minor comments that repeat across a number of files.

File: PollyDemos/DemoBase.cs (hunk 3 above)
Comment (peter-csala) [reply PollyDemos/DemoBase.cs:10]: If you don't mind I would do that renaming as a separate PR after all sync and async demos under the Polly.Demos are migrated.

File: PollyDemos/Async/AsyncDemo05_WaitAndRetryWithExponentialBackoff.cs [hunk 4]
```diff
@@ -38,7 +38,7 @@ public override async Task ExecuteAsync(CancellationToken cancellationToken, IPr
             var strategy = new ResiliencePipelineBuilder().AddRetry(new()
             {
                 ShouldHandle = new PredicateBuilder().Handle<Exception>(),
-                MaxRetryAttempts = 6, // We could also retry indefinitely, but chose six times instead.
+                MaxRetryAttempts = 6, // We could also retry indefinitely by using {whatever the value to do that is}
```
Comment (martincostello) [inline PollyDemos/Async/AsyncDemo05_WaitAndRetryWithExponentialBackoff.cs:41]: Sorry, I didn't mean this to be literal - I meant put whatever the answer is inside the `{}`s 😄

Let me look it up...

File: PollyDemos/Async/AsyncDemo05_WaitAndRetryWithExponentialBackoff.cs [hunk 5]
```diff
@@ -1,112 +1,98 @@
...
-            using (var client = new HttpClient())
+            // Define our strategy:
+            var strategy = new ResiliencePipelineBuilder().AddRetry(new()
             {
-                totalRequests = 0;
-                var internalCancel = false;
-                // Do the following until a key is pressed
-                while (!internalCancel && !cancellationToken.IsCancellationRequested)
+                ShouldHandle = new PredicateBuilder().Handle<Exception>(),
+                MaxRetryAttempts = 6, // We could also retry indefinitely by using {whatever the value to do that is}
```
Comment (martincostello) [inline PollyDemos/Async/AsyncDemo05_WaitAndRetryWithExponentialBackoff.cs:41]:
```suggestion
                MaxRetryAttempts = 6, // We could also retry indefinitely by using int.MaxValue
```

**Review by martincostello — APPROVED** [review]

**Pull request merged**